File: components-registry-service-light-client/src/test/java/org/octopusden/octopus/components/registry/light/client/ComponentsRegistryServiceErrorDecoderTest.java
package org.octopusden.octopus.components.registry.light.client;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import feign.Request;
import feign.Response;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.octopusden.octopus.components.registry.light.client.dto.ErrorResponse;
import org.octopusden.octopus.components.registry.light.client.exceptions.NotFoundException;

import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * The decoder is constructed with a CALLER-supplied ObjectMapper. The server's
 * error body may grow fields (e.g. {@code errorCode}); a strict mapper
 * (FAIL_ON_UNKNOWN_PROPERTIES) must still parse it — otherwise typed 404/425
 * exceptions silently degrade to generic Feign errors. Guarded by
 * {@code @JsonIgnoreProperties(ignoreUnknown = true)} on the DTO.
 */
class ComponentsRegistryServiceErrorDecoderTest {

    private static Response jsonResponse(int status, String body) {
        Map<String, java.util.Collection<String>> headers = new HashMap<>();
        headers.put("content-type", Collections.singletonList("application/json"));
        return Response.builder()
                .status(status)
                .reason("test")
                .request(Request.create(Request.HttpMethod.GET, "/rest/api/2/components/unknown",
                        Collections.emptyMap(), null, StandardCharsets.UTF_8, null))
                .headers(headers)
                .body(body, StandardCharsets.UTF_8)
                .build();
    }

    @Test
    void strictMapperParsesErrorBodyWithUnknownFields() {
        ObjectMapper strictMapper = new ObjectMapper()
                .enable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
        ComponentsRegistryServiceErrorDecoder decoder = new ComponentsRegistryServiceErrorDecoder(strictMapper);

        Exception decoded = decoder.decode(
                "ComponentsRegistryServiceClient#getComponent",
                jsonResponse(404,
                        "{\"errorMessage\":\"Component 'unknown' is not found\","
                                + "\"errorCode\":\"NOT_FOUND\",\"someFutureField\":42}"));

        Assertions.assertInstanceOf(NotFoundException.class, decoded);
        Assertions.assertEquals("Component 'unknown' is not found", decoded.getMessage());
    }

    @Test
    void errorCodeIsExposedAndNullableForOlderServers() throws Exception {
        ObjectMapper strictMapper = new ObjectMapper()
                .enable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);

        ErrorResponse withCode = strictMapper.readValue(
                "{\"errorMessage\":\"uniqueness violation: …\",\"errorCode\":\"UNIQUENESS_VIOLATION\"}",
                ErrorResponse.class);
        Assertions.assertEquals("UNIQUENESS_VIOLATION", withCode.getErrorCode());

        ErrorResponse legacyBody = strictMapper.readValue(
                "{\"errorMessage\":\"plain old error\"}",
                ErrorResponse.class);
        Assertions.assertEquals("plain old error", legacyBody.getErrorMessage());
        Assertions.assertNull(legacyBody.getErrorCode());
    }
}

File: components-registry-service-client/src/test/kotlin/org/octopusden/octopus/components/registry/client/ComponentsRegistryServiceErrorDecoderTest.kt
package org.octopusden.octopus.components.registry.client

import com.fasterxml.jackson.databind.ObjectMapper
import feign.Request
import feign.RequestTemplate
import feign.Response
import feign.RetryableException
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.octopusden.octopus.components.registry.core.exceptions.NotFoundException

class ComponentsRegistryServiceErrorDecoderTest {

    private val decoder = ComponentsRegistryServiceErrorDecoder(ObjectMapper())

    private fun stubResponse(status: Int, reason: String = "", body: String? = null): Response {
        val request = Request.create(
            Request.HttpMethod.GET,
            "http://localhost/test",
            emptyMap(),
            null,
            RequestTemplate()
        )
        val builder = Response.builder()
            .status(status)
            .reason(reason)
            .request(request)
            .headers(emptyMap())
        if (body != null) {
            builder.body(body, Charsets.UTF_8)
            builder.headers(mapOf("content-type" to listOf("application/json")))
        }
        return builder.build()
    }

    @Test
    fun `503 without Retry-After returns RetryableException`() {
        val ex = decoder.decode("test#method", stubResponse(503, "Service Unavailable"))
        assertTrue(ex is RetryableException) { "Expected RetryableException for 503, got ${ex::class.simpleName}" }
        // synthesized by us — no cause
        assertNull(ex.cause)
    }

    @Test
    fun `503 with Retry-After header preserves retry timestamp`() {
        val request = Request.create(
            Request.HttpMethod.GET,
            "http://localhost/test",
            emptyMap(),
            null,
            RequestTemplate()
        )
        val response = Response.builder()
            .status(503)
            .reason("Service Unavailable")
            .request(request)
            .headers(mapOf("Retry-After" to listOf("5")))
            .build()
        val ex = decoder.decode("test#method", response) as RetryableException
        // Verify the exception came from super.decode() (which parses Retry-After),
        // not a freshly synthesized one: super.decode() includes the cause FeignException.
        assertNotNull(ex.cause, "RetryableException from super.decode() should carry the cause FeignException")
    }

    @Test
    fun `404 with json body returns NotFoundException`() {
        val body = """{"errorMessage":"not found"}"""
        val ex = decoder.decode("test#method", stubResponse(404, "Not Found", body))
        assertTrue(ex is NotFoundException) { "Expected NotFoundException for 404, got ${ex::class.simpleName}" }
    }

    @Test
    fun `500 returns generic FeignException`() {
        val ex = decoder.decode("test#method", stubResponse(500, "Internal Server Error"))
        assertTrue(ex is feign.FeignException) { "Expected FeignException for 500, got ${ex::class.simpleName}" }
        assertTrue(ex !is RetryableException) { "500 should not be retryable" }
    }
}
